File: leetcode/twosum.c
/**
 * @file twosum.c
 * @author ythatshashanlguy 
 * @brief Just Writing few Leetcode questions to get more hands on C 
 * @version 0.1
 * @date 2022-09-13
 * 
 * @copyright Copyright (c) 2022
 * 
 */

/**
 * 
 * @brief Problem Statement 
 * Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
 * You may assume that each input would have exactly one solution, and you may not use the same element twice.
 * You can return the answer in any order.
 * @note the return array must be malloced , assume caller calls free();
 */

#include <stdlib.h>
#include<stdio.h>

//brute force approach

int* twoSum(int* nums, int numsSize, int target, int *returnSize ){
    int i , j;
    int * output_arr = (int*)malloc(2*sizeof(int));

    if(output_arr == NULL || nums == NULL){
        *returnSize  ==0;
        return NULL;
    }

    for (i = 0; i < numsSize; i++){
        for(j = i+1;j < numsSize;j++){
            if(nums[i] + nums[j] == target)
            *returnSize = 2; 
            output_arr[0] = j ;
            output_arr[1] = i;
            return output_arr;
        }
    }

    *returnSize = 0 ;
    free(output_arr);

    return NULL;

    
}

/**
 * @brief optimizing the two sum funciton by using hash table 
 * 
 * @param nums 
 * @param numsSize 
 * @param target 
 * @param returnSize 
 * @return int* 
 */
int* twoSum(int* nums, int numsSize, int target, int *returnSize ){
// to be impelemented once Hash table is clear ; 
}
int main(){
    int a[] = {2,7,11,15};
    int rs,target = 9;
    int *p = NULL;

    p = twoSum(a,sizeof(a)/sizeof(a[0]),target,&rs);

    if(rs== 0 ){
        printf("target not found ");
    }else{
        printf("target found");
    }
}

File: leetcode/twosum.js

/**
 * 
 * @brief Problem Statement 
 * Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
 * You may assume that each input would have exactly one solution, and you may not use the same element twice.
 * You can return the answer in any order.
 * @note the return array must be malloced , assume caller calls free();
 */

arr = [2,7,11,15];
target = 9 ; 

/**
 * 
 * @param {*} arr 
 * @param {*} t 
 */
function twosum_brute(arr,t) {
    for(let i = 0 ; i <arr.length ;i++){
        for(let j = i+ 1 ; j< arr.length ; j++){
            if(target === arr[i] + arr[j]){
                    return [i,j]
            }
        }
    }
    return[]
}

/**
 * Optimize method 
 * @param {*} arr 
 * @param {*} t 
 * @returns 
 */

function twoSum(arr,t){
     const hash_table = {}
     for(let i = 0 ; i < arr.length ; i++){

        if(hash_table[t - arr[i]] !== undefined){
                return[hash_table[t-arr[i]],i]
        }else{
            hash_table[arr[i]] =  i ;
        }
     }
     return []
}


/**
 * Calling both optimal and brute force approch and timing the to understand time complexity better 
 * results 
 * [ 0, 1 ]
 * Brute Force : 8.078ms
 *  [ 0, 1 ]
 *  Optimal : 0.168ms
 * 
 */
console.time('Brute Force ')
console.log(twoSum(arr,target))
console.timeEnd('Brute Force ')


console.time('Optimal ')
console.log(twosum_brute(arr,target))
console.timeEnd('Optimal ')